analysis.py

```diff
@@ -0,0 +1,211 @@
+            print("this line doesn't contain valid addr and app_pw: %s" %
+                  (entry["line"],))
+
+    # create test group. who is in it after 60 seconds?
+    begin = time.time()
+    print("Creating group " + str(begin))
+    group = spac.create_group_chat("Test Group " + str(begin),
+                                   contacts=[spac.create_contact(entry["addr"]) for entry in credentials])
+    group.send_text("Sender: spider\nBegin: %s" % (str(begin),))
+    group_members = []
+    while time.time() < float(begin) + args.timeout:
+        if len(accounts) == len(group_members):
+            break
+        for ac in accounts:
+            if ac in group_members:
+                continue
+            for chat in ac.get_chats():
+                if chat.get_name() == group.get_name():
+                    group_members.append(ac)
+                    output.submit_groupadd_result(ac.get_self_contact().addr, time.time() - begin)
+    else:
+        print("Timeout reached. Not added to group: ", end="")
+        for ac in accounts:
+            if ac not in group_members:
+                print(ac.get_self_contact().addr, end=", ")
+        print()
+
+    begin = time.time()
+    while time.time() < float(begin) + args.timeout:
+        if group_members == []:
+            break
+        for ac in group_members:
+            for chat in ac.get_chats():
+                if chat.get_name() == group.get_name():
+                    grp = chat
+            msgs = grp.get_messages()
+            for msg in msgs:
+                msgcontent = parse_msg(msg.text)
+                if msg.time_received is None or msgcontent.get("sender") == "spider":
+                    continue
+                msgreceived = (msg.time_received - datetime.datetime(1970, 1, 1)).total_seconds()
+                duration = msgreceived - msgcontent.get("begin")
+                print("%s received message from %s after %.1f seconds" % (ac.get_self_contact().addr, msgcontent["sender"], duration))
+                output.submit_groupmsg_result(ac.get_self_contact().addr, msgcontent["sender"], duration)
+                group_members.remove(ac)
+
+    # send test messages with spider
+    for ac in accounts:
+        if spider is not None:
+            if spider["addr"] is not ac.get_self_contact().addr:
+                check_account_with_spider(spac, ac)
+                ac.begin = time.time()
+
+    # wait until finished, or timeout
+    while len(accounts) > 0:
+        for ac in accounts:
+            if ac._shutdown_event.is_set():
+                accounts.remove(ac)
+            elif time.time() > ac.begin + args.timeout:
+                print("%d seconds timeout while waiting for echo to %s - test failed." %
+                      (args.timeout, ac.get_self_contact().addr))
+                output.submit_1on1_result(ac.get_self_contact().addr, "timeout", "timeout")
+                ac.shutdown()
+                ac.wait_shutdown()
```

Comment:
This is a bit hacky. during the loop all of the accounts are shut down, and then slowly removed. But how important is it to safely shutdown the accounts? When the script ends, all of those are killed anyway.

This timeout loop is entangled closely to the last test; this makes it really difficult to just move this test somewhere else, or change it fundamentally, if you want to keep the timeout as well. This is suboptimal and hacky; maybe checking if the test succeeded should be separated from shutting down the accounts.

On the other hand it's fun ;)

Reply:
fixed in #5 :)